### DISCUSSION TITLE
Allow assigning nullable enum class member that has the same name as enum.

### DISCUSSION THREAD

**Discussion by edward-a** [dotnet/csharplang#6850] [Language Ideas]
Consider a nullable enum variable with the same name as enum:

` enum E
  {
      V
  }
`
 
` class C
  {
      E? E;
...
`

The below assignment in the class constructor will not compile:

    C() {
      E = E.V;
    }
  }

With a regular (non-nullable) enum it compiles just fine. The proposal is to make such assignment work for nullable enums to make them as convenient to use as the regular ones.

**Comment by 333fred** [comment]
This is likely just a bug. Please file it on the Roslyn repository, following the template for a bug report.

**Reply by CyrusNajmabadi** [reply]
I think this is technically by design. This is the `Color Color`. Except here it's not `Color`, it's `Nullable<Color>`.  I think we would need alamg change here technically.  Though I'd support it (esp as I think this would work for reference types, just not value types).

**Reply by HaloFour** [reply]
I could see where it would be ambiguous, especially if the enum had a member named `Value`.  Would that be `E.Value` or `Nullable<E>.Value`?

**Reply by CyrusNajmabadi** [reply]
Good point!

**Reply by HaloFour** [reply]
It's probably such a narrow case that it's worth breaking if the ergonomic benefits are considered worthwhile.  🤷

**Reply by bernd5** [reply]
Not sure if it is a bug - `E` can mean really both here, the type or the field (maybe we have some special binding rules...).

To let it work you can use a using-alias or a full-qualified-type name:
```csharp
new C();

enum E { V }
class C{
    E? E;
    public C() {
      E = global::E.V;
    }
}
```

In this case - because it is an enum the non-static member are very limited. But there are still some instance members, take for example:
```csharp
new C();

enum E { 
    ToString
}

class C{
    E? E;
    public C() {
      E = E.ToString;
    }
}
```
What is meant here:
- the `ToString` method group from `Nullable<E>`
- the `ToString` method group from `E`
- the enum member `ToString`?

Currently roslyn issues: `error CS0428: Cannot convert method group 'ToString' to non-delegate type 'E?'. Did you intend to invoke the method?`

Btw.: how could I achieve the following:
```csharp
using System;

new C();

enum E { 
    ToString
}

class C{
    E E;
    public C() {
      E = E.ToString;
      Func<string> s = E.ToString; //<-- currently `E.ToString` binds to the enum member not the method group
    }
}
```

**Reply by bernd5** [reply]
Btw. it is not a special case for `enum` only:
```csharp
using System;

new C();

struct E1 { 
    public static int V;
    
    public static implicit operator E1(int i){
        return new E1();
    }
}

enum E2 { V }
class C{
    E1? E1;
    E2? E2;
    public C() {
      E1 = E1.V;
      E2 = E2.V;
    }
}
```

### Roslyn Diagnostics:   

- > error CS1061: 'E1?' does not contain a definition for 'V' and no accessible extension method 'V' accepting a first argument of type 'E1?' could be found (are you missing a using directive or an assembly reference?)

- > error CS1061: 'E2?' does not contain a definition for 'V' and no accessible extension method 'V' accepting a first argument of type 'E2?' could be found (are you missing a using directive or an assembly reference?)

**Reply by 333fred** [reply]
Ok, I dug into the spec a bit: https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#11762-identical-simple-names-and-type-names is the relevant section. Specifically, this quote (emphasis mine):

> In a member access of the form `E.I`, if `E` is a single identifier, and if the meaning of `E` as a *simple_name* ([§11.7.4](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1174-simple-names)) is a constant, field, property, local variable, or parameter **with the same type as the meaning of `E` as a *type_name* ([§7.8.1](basic-concepts.md#781-general))**, then both possible meanings of `E` are permitted.

Since `E` here is a property with the type `E?`, and that is not the same as the type `E`, these special lookup rules don't apply and the compiler is behaving according to spec. It does feel like something we could address with only a minor chance at a breaking change (the previously mentioned `Value` case).

**Reply by CyrusNajmabadi** [reply]
@333fred Agreed. I'm amenable to it.  It feels unfortunate that just saying the prop can be nullable turns off `Color Color`.    But i also don't feel strongly about this as important to change.

**Reply by edward-a** [reply]
> > In a member access of the form `E.I`, if `E` is a single identifier, and if the meaning of `E` as a _simple_name_ ([§11.7.4](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1174-simple-names)) is a constant, field, property, local variable, or parameter **with the same type as the meaning of `E` as a _type_name_ ([§7.8.1](basic-concepts.md#781-general))**, then both possible meanings of `E` are permitted.

I am wondering if the rule can be extended as follows:

"... with the type **assignable from `E`** meant as a _type_name_ ([§7.8.1](basic-concepts.md#781-general))"

**Reply by 333fred** [reply]
I think that's far too broad. It should just special case nullable here.

**Comment by svick** [comment]
> I think this would work for reference types, just not value types

[It seems it already does work for nullable reference types:](https://sharplab.io/#v2:CYLg1APgAgTABAUQLACgDeq5blAzDgRgDZE4BlAbk2z1IEkqUBfVVWOAYVQxW0QH5EFOK17YOACgCU1LDz58EcALyIAdJVkKlqhGoZaWzIA=)

```c#
class E
{
    public static E S;
    public E I;
}

class C
{
    E? E; 

    C()
    {
        E = E.S;
        E = E.I;
    }
}
```

The above code compiles fine.

Not sure if that's how it should behave, according to the spec.

**Reply by CyrusNajmabadi** [reply]
it should work.  For reference types, the name of the type is unchanged by `?` so the `Color Color` case still holds.  it's just interesting that `?` with a struct makes the type `Nullable<T>` from `T`, making this no longer apply.

**Reply by 333fred** [reply]
@svick see my comment [here](https://github.com/dotnet/csharplang/discussions/6850#discussioncomment-4584805).
